### DISCUSSION TITLE
Weird behavior when reading an ADC chip

### DISCUSSION THREAD

**Discussion by thedeany** [FortySevenEffects/arduino_midi_library#311] [Q&A]
I've been trying to troubleshoot this for quite a while and am about at my wits' end, so thank you in advance to any help you can give!

I'm trying to create a small interface of sorts to communicate to and from VCV Rack software to my physical modular synth. I'm attempting to do so via an MCP4728 4-channel DAC and an ADS1115 4-channel ADC. I have both of them hooked up to the SCL/SDA pins on my Arduino Uno and they are using different addresses, and the chips seem to operate as one would expect.

I have a simple LFO set up in VCV Rack and am outputting that via a CV->MIDI CC module to test with. I'm then running [serialmidi.py](https://github.com/raspy135/serialmidi) to create the serial-to-MIDI bridge.

When I run the following code **_with `readADC();` in `loop()` commented out_**, I can see the MIDI values (via MIDI Monitor, monitoring channel 5 which I've set up as kind of a reflector) incrementing appropriately, and the correct voltage is output from the DAC (per my multimeter.)

However, when I run the code **_as written_**, it works okay for about 15 seconds or so and then I start seeing `48` as my MIDI value every other or every few messages in the monitor, rendering the entire thing useless. It's always `48`. I've searched for significance with that number, but nothing comes up.

I've been thinking that maybe my issue is there's too much data on the I2C bus with both of these chips being active/read. Could that actually be the issue here? If so, any recommendations? If not, what else am I missing?

```cpp
#include <Adafruit_MCP4728.h>
#include <MIDI.h>
#include <Adafruit_ADS1X15.h>

Adafruit_ADS1115 ads;

MIDI_CREATE_DEFAULT_INSTANCE();

Adafruit_MCP4728 mcp;
int channels[4] = { 0, 1, 2, 3 };

void setup(void)
{
  MIDI.begin(MIDI_CHANNEL_OMNI);
  MIDI.turnThruOff();
  MIDI.setHandleControlChange(handleCC);

  // Try to initialize!
  if (!mcp.begin(0x64)) {
    while (1);
  }

  if (!ads.begin(0x48)) {
    while (1);
  }
}

void handleCC(byte channel, byte control, byte value)
{
  if (channel == 1)
  {
    long mappedValue = map(value, 0, 127, 0, 4095);
    mcp.setChannelValue(channels[control], mappedValue);

    // Send thru on different channel for debugging purposes
    MIDI.sendControlChange(control, value, 5);
  }
}

void readADC()
{
  int16_t adc0;

  adc0 = ads.readADC_SingleEnded(0);

  long value = map(adc0, 0, 26600, 0, 127);
  MIDI.sendControlChange(2, value, 2);
}

void loop()
{
  readADC();
  MIDI.read();
  delay(5);
}
```

**Comment by franky47** [comment]
The value 48 comes from the output of `map(adc0, 0, 26600, 0, 127)`, if you reverse that mapping, would the output of the ADC have any particular meaning?

**Reply by thedeany** [reply]
The ADC will eventually be meant to take in a voltage (the 26600 max value I have here is roughly 5V), map it to MIDI range, and send the MIDI value to VCV Rack for use there. So I don't think the mapping can be reversed. Unless I'm misunderstanding what you're asking.

**Reply by franky47** [reply]
Some other possible pointers:
https://github.com/adafruit/Adafruit_ADS1X15/issues/14

**Reply by franky47** [reply]
> The ADC will eventually be meant to take in a voltage (the 26600 max value I have here is roughly 5V), map it to MIDI range, and send the MIDI value to VCV Rack for use there. So I don't think the mapping can be reversed. Unless I'm misunderstanding what you're asking.

I meant that whatever value the ADC returns gets mapped into 48. "reversing" the map would be calculating that input value (the value read by the ADC) and see if it has a particular significance in the chip manual.

**Reply by thedeany** [reply]
Ugh, I think this might be hardware misconfiguration on my end. None of the examples I saw for either of these chips used pull-up resistors, but I'm now reading in the datasheets for both that they need them on the SCL and SDA lines 😩

I'll have to give that a shot tomorrow. Thanks for the quick reply. I'll reply here with whichever way the pull-up resistor test goes.

**Reply by franky47** [reply]
IIRC there are built-in pull-ups in AVR chips, you could look into the ATmega328P datasheet how to set those up and see if they are enough to drive the I2C bus.

**Reply by thedeany** [reply]
Alright, I've had some time to get back to test the pull-ups. From what I've found online, it seems the internal pull-ups are anywhere from 20k-50k which seems a bit high. It was surprisingly difficult to find out how to disable them, but I believe
```
digitalWrite(A4, LOW);
digitalWrite(A5, LOW);
```
does it on the Uno. I've added those commands and added a 4.7k pull-ups from the SCL and SDA lines to 5V.

This has unfortunately had zero effect; I'm still seeing the errant messages about 15-18 seconds after booting.

**Reply by thedeany** [reply]
Okay, another piece of the puzzle.

This whole time, I've been testing with a sawtooth LFO. This results in MIDI messages being output, at the LFO rate I had set, about once every 100ms. I just now switched to a quite slow square wave LFO, about 3s between each MIDI message, and let it run for a couple minutes and never saw the `48` message once. I then started to increase the frequency of the square wave LFO and sure enough when the MIDI messages got to about 100ms apart, the `48` message started coming up.

So now instead of an I2C issue, I'm thinking that something (whether that be the Arduino, my code, the serialmidi.py bridge script, or the library itself) is not handling this rate of messages well.

**Comment by Grumpy-Mike** [comment]
>  It was surprisingly difficult to find out how to disable them,

To remove the internal pull up resistors on a Uno's I2C lines use the following:-
```
pinMode(A4, INPUT);
pinMode(A5, INPUT);
```
